### ISSUE TITLE
Idea: abstract out vectorizer_model.get_feature_names_out() out and make words reusable

### ISSUE THREAD

**Issue by zxygentoo** [MaartenGr/BERTopic#1260]
Hi @MaartenGr 

Continuing my work mentioned in: https://github.com/MaartenGr/BERTopic/issues/1138

Using these ML tokenizor is relatively slow specially on large datasets. For my test, on a dataset consists of half million sentences, after clustering it will take around one hour(paddle as tokenizor on GPU) to finish which is the most time consuming step right now. I guess this issue exists for Japanese, Korean and many other languages.

A quick skim at the code, seems the `words = vectorizer_model.get_feature_names_out()` parts should be able to be abstracted out? Then we can store the result somewhere and reuse it in subsequent runs just like the pre-computed embeddings?

Just like to get your opinion on this. If this is feasible and desirable, I may take a crack on it.

Cheers,
Alex

**Comment by zxygentoo** [comment]
Or this is just my ignorance of sklearn and this is already supported on their side?

Right now I use some this like this

```python
embeddings = sentence_model.encode(sentences, show_progress_bar=True, convert_to_numpy=True)
embeddings = embeddings / np.linalg.norm(embeddings, axis=1, keepdims=True)
with open(save_to, "wb") as f:
    pickle.dump({'sentences': sentences, 'embeddings': embeddings}, f)
```

to save the encoded embeddings to disk and load it to reuse in multiple runs.

Wondering if the same thing is possible for CountVectorizer?

Any pointers? Thanks~

**Comment by MaartenGr** [comment]
Just to make sure I understand correctly. The tokenizer that you created in the issue that you mentioned is already working, right? So why would there be a need to abstract out `words = vectorizer_model.get_feature_names_out()`?

**Comment by zxygentoo** [comment]
Yes it's all working. But the word tokenization part take quite some time to finish each run.

So the idea is (not sure if it's feasible) if we could tokenize all the sentences once and store the result to disk, will we be able to reuse them later(like store/loading encoded sentence embbdings)? Word segmentation for language like Chinese(specially if using some machine learning driven methods for accuracy) is pretty slow comparing to languages like English.

**Comment by MaartenGr** [comment]
You could create a custom class and make it so that `.transform` returns a pre-calculated bag-of-words and that the `.fit` simply returns itself. So something like this:

```python
class CustomCV:
    def __init__(self, bow, words):
        self.bow = bow
        self.words = words
        
    def transform(self, X):
        if len(X) != self.bow.shape[0]:
            return csr_matrix(np.ones((len(X), bow.shape[1])))
        else:
            return bow
    
    def fit(self, X, y=None):
        return self
    
    def get_feature_names_out(self):
        return self.words
```

However, I cannot say for certain that this will work since the vectorizer model is also used to extract representative documents and a number of other functions. Moreover, the bag-of-words (`bow`) is not a bag-of-words per document but per topic so you would first have to run your topic model and then, again, need to calculate the `bow` which can be saved:

```python
from sklearn.datasets import fetch_20newsgroups
from bertopic import BERTopic
from umap import UMAP
import pandas as pd

# Data
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

# Train a BERTopic model
umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', random_state=42)
topic_model = BERTopic(umap_model=umap_model).fit(docs)

# Calculate the `bow` and the `words` that are necessary
documents = pd.DataFrame({"Document": docs,
                          "ID": range(len(docs)),
                          "Topic": topic_model.topics_})
documents_per_topic = documents.groupby(['Topic'], as_index=False).agg({'Document': ' '.join})
documents_per_topic = topic_model._preprocess_text(documents_per_topic.Document.values)
bow = topic_model.vectorizer_model.transform(documents_per_topic)
words = topic_model.vectorizer_model.get_feature_names_out()

# Use our `CustomCV` to skip over calculation of `bow`
vectorizer_model = CustomCV(bow=bow, words=words)
umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', random_state=42)
topic_model = BERTopic(umap_model=umap_model, vectorizer_model=vectorizer_model).fit(docs)
```

Just tested this and it should work although many features, such as `topics_over_time` that rely on `vectorizer_model` will not work as they re-calculate `bow` based on specific slices.

**Comment by zxygentoo** [comment]
This is what I'm worrying about (should've read the code more carefully). As the input of `vectorizer_model` is `documents_per_topic`, it essentially changes every time we change the parameters of `umap_model` and `hdbscan_model`.

I's clear to me now: the API of BERTopic doesn't prevent this sort of thing, but it will take a lot more effect than was imaged at the `VectorizeCount` side: eg. maybe make a custom analyzer to directly use pre-tokenized corpus in its `transform` then do some doc id matching in a custom `vectorizer_model` at BERTopic side each time after clustering. Then maybe I can tokenize all the docs once  and store/reload as long as the docs and tokenizer remains the same(which is the typical case as the main parameters need to be tuned are HDBScan/UMAP). The way sklearn is written has pretty deep class hierarchy, this may take some time to work out.

Thanks for the patient response, this helps a lot!

**Comment by MaartenGr** [comment]
Although the IDs are not passed to the `vectorizer_model`, you could try to match the documents against one another and derive the IDs from that matching procedure. It would require some custom CountVectorizer but it might be feasible.
